samza-core/src/main/java/org/apache/samza/clustermanager/ContainerManager.java

```diff
@@ -80,19 +82,23 @@
 
   private final Optional<StandbyContainerManager> standbyContainerManager;
 
+  private final LocalityManager localityManager;
```

Comment:
Now we are passing this LocalityManager everywhere ContainerManager, CPM, StandbyContainerManager just to get the host that container was last seen....

I think we can have it cleaner by either maintaining that in SamzaApplicationState or having static util in LocalityManager to fetch the last seen host

Reply:
1. `SamzaApplicationState` is a holder class to share state between components and not supposed to hold on any manager classes. We have a TODO to remove JobModelManager out of it. It would be counter productive to introduce another component into the holder class. 
2. I am not a huge fan of Static for following reasons

- It makes testing harder and using power mock for component testing everywhere.
- Utility classes that handles components with lifecycle is harder to reason about.
- Static methods are harder to evolve without breaking the signature & modifying the callers

Reply:
+1, using state to fetch locality information was a bad pattern.
Makes writing tests super weird.